README.md
# Sitespeed Grade

Sitespeed 性能评级指标

## 安装使用

```shell
npm install sitespeed-grade
npm install sitespeed.io
npx sitespeed.io https://github.com --plugins.add ./node_modules/sitespeed-grade/lib/index.js
```

集成 Grafana 时，如果需要在 Grafana 上显性能评级指标，需要按以下几点要求配置

1. 将官方的 sitespeed/grafana-bootstrap 替换为 zhbhun/grafana-bootstrap。
2. sitespeed.io 执行时先移除 "graphite"，再重新添加 "graphite"（保证 sitespeed-grade 插件在 graphite 之前执行）

    ```shell
    npx sitespeed.io https://github.com --plugin.remove graphite --plugins.add ./node_modules/sitespeed-grade/lib/index.js --plugin.add graphite
    ```
## 开发调试

```shell
docker-compose pull
mkdir volumes
mkdir volumes/grafana
chmod 777 volumes/grafana
mkdir volumes/whisper
docker-compose up -d
npm run debug # 调试
npm run test:latency # 测试网络延迟对性能的影响
npm run test:download # 测试下载速率对性能的影响
npm run test:grade # 测试常见网页的性能测试结果
# 访问 http://localhost:3000 查看 grafana 数据
# 修改 lib/defaultConfig 调整性能评级规则
```

## 配置说明

```js
{
  grade: {
    total: 10, // 评级总分，默认为 10
    metrics: [ // 评级指标集合
      {
        name: 'xxx', // 指标名称,
        weight: 10, // 指标评级权重
        negative: false, // 是否计入总权重
        targets: function (value) {
          // 计算返回得分率
        },
        value: 'browserScripts.${iteration}.timings.pageTimings.redirectionTime' // 原始值（直接获取）
      },
      {
        name: 'xxx', // 指标名称,
        weight: 10, // 指标评级权重
        negative: false, // 是否计入总权重
        targets: [ // 得分目标值的计算因子(不区分网络)
          [100, 1], 
          [200, 0.9],
          [300, 0.8],
          [400, 0.7],
          [600, 0.5],
          [1000, 0.3]
        ],
        value: 'browserScripts.${iteration}.timings.pageTimings.redirectionTime' // 原始值（直接获取）
      },
      {
        name: 'xxx', // 指标名称,
        weight: 10, // 指标评级权重
        targets: { // 得分目标值的计算因子，key 值为网络配置的别名，value 值时计算因子的配置项
          '3g': [ // 3G 网络得分目标值区间
            [100, 1], 
            [200, 0.9],
            [300, 0.8],
            [400, 0.7],
            [600, 0.5],
            [1000, 0.3]
          ],
          '4g': [] // 4G 网络得分目标值区间
        },
        value: 'browserScripts.${iteration}.timings.pageTimings.redirectionTime' // 原始值（直接获取）
      },
      {
        name: 'xxx', // 指标名称,
        weight: 10, // 指标评级权重
        targets: { // 得分目标值的计算因子
          '3g': 300, // 3G 网络得分目标值
          '4g': 100 // 4G 网络得分目标值
        },
        value: [
          'browserScripts.${iteration}.timings.navigationTiming.responseStart',
          'browserScripts.${iteration}.timings.navigationTiming.requestStart'
        ] // 原始值（差值计算）
      },
    ]
  }
}
```

计算原理如下所示

1. 计算指标项的原始值（根据 value 计算获得）；

    如果原始值不存在，则忽略该项指标

2. 根据指标项的目标值计算得分因子（用上一步获得的原始值和目标值 targets 比对）；
3. 计算指标项总权重（只有 negative 不是 true 的才算在内）

    如果得分因子为小于 0，可以认为该项指标作为附件扣分项，不计入总权重

3. 计算评级得分：评级总分 * ∑(指标项得分因子 * 指标项权重 / 指标项总权重 )

### 默认配置

可用的性能指标

1. 技术指标

    - 网页传输大小
    - 重定向时间
    - 网络连接时间
    - 服务端响应时间
    - 服务端下载时间

2. 以用户为中心的指标

    - 首次绘制
    - 首次内容绘制
    - 首次有效绘制
    - 可交互时间
    - 速度指数

默认配置参考 [defaultConfig.js](./lib/defaultConfig.js)

### 网络配置

根据最新的 [中国宽带速率状况报告_第23期（2019Q1）.pdf](http://www.chinabda.cn/Site/Default/Uploads/kindeditor/file/20190522/%E4%B8%AD%E5%9B%BD%E5%AE%BD%E5%B8%A6%E9%80%9F%E7%8E%87%E7%8A%B6%E5%86%B5%E6%8A%A5%E5%91%8A_%E7%AC%AC23%E6%9C%9F%EF%BC%882019Q1%EF%BC%89.pdf)，可知目前 4G 的平均速率在 23.01Mbit/s 左右，3G 的平均速率在 8.89Mbit/s 左右。由于移动宽带网络由基站覆盖范围内的大量用户共享，区域内的人员密集程度会直接影响到每个用户所能体验到的下载速率。另一方面，移动宽带用户上网体验速率还会受到建筑物阻隔、用户的位置移动速率、用户终端配置等影响。所以在模拟测试的时候，建议在统计出来的数据基础上再打个半折。

- 4G

    - downstream：12288Kbps / 12Mbps
    - upstream: 3072Kbps / 3MBps
    - rtt: 30

- 3G

    - downstream：4096Kbps / 4Mbps
    - upstream: 1024Kbps / 1MBps
    - rtt: 50

## 测试数据

### 延迟测试

ps：测试脚本采用了 [tsproxy](https://github.com/WPO-Foundation/tsproxy) 来模拟延迟，上下载的速率均为 9216 Kbps。并且设置了一组不同的 latency（0、200、300、400、500、600、700、800、900），通过测试结果分析 latency 值对网页加载的影响。

运行脚本

```bash
npm run test:latency
```

测试结果

```bash
baidu 0
 connect - median: 88 | mean: 85 | min: 76 | max: 92
 response - median: 261 | mean: 269 | min: 230 | max: 315
 download - median: 147 | mean: 148 | min: 115 | max: 183
baidu 200
 connect - median: 495 | mean: 570 | min: 470 | max: 746
 response - median: 550 | mean: 542 | min: 457 | max: 620
 download - median: 183 | mean: 177 | min: 162 | max: 187
baidu 300
 connect - median: 693 | mean: 796 | min: 676 | max: 1019
 response - median: 622 | mean: 617 | min: 599 | max: 630
 download - median: 159 | mean: 161 | min: 133 | max: 191
baidu 400
 connect - median: 898 | mean: 1018 | min: 880 | max: 1275
 response - median: 659 | mean: 652 | min: 585 | max: 711
 download - median: 160 | mean: 135 | min: 82 | max: 162
baidu 500
 connect - median: 1160 | mean: 1279 | min: 1090 | max: 1588
 response - median: 724 | mean: 730 | min: 709 | max: 758
 download - median: 99 | mean: 105 | min: 73 | max: 143
baidu 600
 connect - median: 1284 | mean: 1483 | min: 1276 | max: 1890
 response - median: 843 | mean: 861 | min: 834 | max: 905
 download - median: 137 | mean: 132 | min: 110 | max: 148
baidu 700
 connect - median: 1468 | mean: 1707 | min: 1467 | max: 2185
 response - median: 922 | mean: 937 | min: 888 | max: 1001
 download - median: 107 | mean: 119 | min: 99 | max: 150
baidu 800
 connect - median: 1688 | mean: 1946 | min: 1671 | max: 2479
 response - median: 1109 | mean: 1106 | min: 1054 | max: 1154
 download - median: 126 | mean: 128 | min: 106 | max: 152
baidu 900
 connect - median: 1896 | mean: 2182 | min: 1853 | max: 2796
 response - median: 1163 | mean: 1157 | min: 1113 | max: 1194
 download - median: 147 | mean: 126 | min: 82 | max: 149
```

分析

- 网络连接：2 * latency + 原有延迟时长
- 响应：latency + 原有响应时长
- 下载：不影响

### 下载速度测试

ps：测试脚本采用了 [tsproxy](https://github.com/WPO-Foundation/tsproxy) 来模拟延迟，延迟均为 0。并且设置了一组不同的下载速度（1024、2048...9120），上传速度为四分之一的下载速度，通过测试结果分析下载速率值对网页加载的影响。

运行脚本

```bash
npm run test:latency
```

```bash
baidu 1024
 connect - median: 86 | mean: 91 | min: 74 | max: 114
 response - median: 811 | mean: 847 | min: 808 | max: 922
 download - median: 674 | mean: 677 | min: 655 | max: 701
baidu 2048
 connect - median: 97 | mean: 101 | min: 82 | max: 123
 response - median: 469 | mean: 475 | min: 460 | max: 496
 download - median: 325 | mean: 327 | min: 320 | max: 337
baidu 3072
 connect - median: 63 | mean: 75 | min: 59 | max: 102
 response - median: 368 | mean: 360 | min: 336 | max: 377
 download - median: 221 | mean: 223 | min: 214 | max: 235
baidu 4096
 connect - median: 82 | mean: 86 | min: 78 | max: 98
 response - median: 288 | mean: 335 | min: 278 | max: 438
 download - median: 174 | mean: 176 | min: 164 | max: 189
baidu 5120
 connect - median: 88 | mean: 92 | min: 83 | max: 105
 response - median: 430 | mean: 424 | min: 290 | max: 552
 download - median: 282 | mean: 249 | min: 135 | max: 329
baidu 6144
 connect - median: 93 | mean: 88 | min: 77 | max: 94
 response - median: 295 | mean: 302 | min: 259 | max: 352
 download - median: 180 | mean: 182 | min: 112 | max: 254
baidu 7168
 connect - median: 96 | mean: 95 | min: 83 | max: 105
 response - median: 368 | mean: 326 | min: 240 | max: 369
 download - median: 164 | mean: 166 | min: 107 | max: 226
baidu 8192
 connect - median: 65 | mean: 63 | min: 56 | max: 68
 response - median: 252 | mean: 241 | min: 212 | max: 259
 download - median: 113 | mean: 105 | min: 88 | max: 114
baidu 9216
 connect - median: 171 | mean: 160 | min: 123 | max: 185
 response - median: 324 | mean: 329 | min: 248 | max: 416
 download - median: 150 | mean: 138 | min: 104 | max: 160
```

- 网络连接：不受影响
- 响应：在上传速度较低时会影响响应速度
- 下载：根据传输文件大小和下载速率有关

    例子中百度移动端首页大小大于在 90KB 左右，在测试例子中下载速度从 1024 提升到 4096 之间，都有缩小相应倍数的下载时长，但继续提升下载速度时，影响不是很大了。

    可以根据计算公式 `（90KB * 8bit / 下载速度）` 粗略计算下载时长：

    - 90 * 8 / 1024 = 0.703125s
    - 90 * 8 / 2048 = 0.3515625s
    - 90 * 8 / 3172 = 0.22698612862547288s
    - ...

### 首屏体验测试

参见 [paint](./test/paint/README.md)

## 误区

- 性能得分低，代表性能优化的不好

    得分低并不代表性能优化的不好，可能与具体的业务有关系，例如：电商类的业务资源比较多，加载性能得分自然比较低。

- 业务本身就复杂，已经尽力优化了

    除了从前端代码层面的技术方向上去优化加载性能，也可以考虑与客户端合作，通过预加载和离线包的方式去优化。

- ...

## 原理

有两大因素影响着网页加载性能，一是网络，二是网页本身实现的优化情况。而我们性能测试得分的主要目标是要在相同网络环境下，检测不同网页的加载性能情况。在得出网页性能扽分之前，需要明白网络是怎样影响加载性能的，然后在固定网络的情况下测试得出性能指标值，最终计算得出一个相对稳定可靠的性能得分。

1. 网络延迟对网页加载性能的影响
2. 网络速度对网页加载性能的影响
3. 首屏时间对用户体验的影响
4. 基于调研结果开发性能指标的得分计算函数

    需要借助一些数学知识，通过非线性的曲线函数去计算不同指标值的得分

5. 综合测试常用网站的网络性能指标，并取舍和配置不同性能指标的权重，最终得出一个唯一的性能得分


CHANGELOG.md
# Chnagelog

## 1.2.0

- 通过测试首屏时间对用户体验的影响来调整指标得分配置

    提高了首屏时间和首次有效绘制时间的要求

- 通过曲线函数来计算指标得分值的精准和契合度
- 补充测试数据文档和性能评分的开发思考

## 1.1.0

- 增加测试脚本
- 调整性能得分的默认配置（以用户体验为中心）


test/grade.js
/**
 * 综合测试
 */
const path = require('path');
const fs = require('fs');
const fse = require('fs-extra');
const shell = require('shelljs');
const chalk = require('chalk');
const config = require('../.sitespeedio.test.json');
const SitespeedGrade = require('../lib');
const findDataPath = require('./utils/findDataPath');
const websites = require('./websites.json');

const context = path.resolve(__dirname, '../');
const output = path.resolve(context, 'output/grade');

function test({ type, name, page, url }) {
  return new Promise(async (resolve, reject) => {
    const outoutPath = path.resolve(output, type, name, page);
    const cmd = `sitespeed.io "${url}" --config ./.sitespeedio.test.json --outputFolder ${outoutPath}`;
    console.log('$', chalk.blue(cmd));
    if (await fse.exists(outoutPath)) {
      reject(new Error('aready run.'));
      return;
    }
    await fse.ensureDirSync(outoutPath);
    await fse.emptyDir(outoutPath);
    shell.exec(cmd, {
      cwd: context,
      async: true,
      fatal: false,
      silent: false,
    }, async (code, stdout, stderr) => {
      if (code === 0) {
        const pagesPath = path.resolve(outoutPath, 'pages');
        await fse.copy(
          await findDataPath(pagesPath),
          outoutPath,
          { overwrite: true }
        );
        await fse.emptyDir(pagesPath);
        resolve(stdout);
      } else {
        reject(new Error(stderr));
      }
    });
  });
}

async function runWebPage(website, index) {
  if (website.pages && website.pages[index]) {
    const webpage = website.pages[index];
    try {
      await test({
        type: website.type,
        name: website.name,
        page: webpage.name,
        url: webpage.url,
      });
      console.log(chalk.green('success\n'));
    } catch (error) {
      console.log(chalk.red(`failure: ${error.message}\n`));
    }
    await runWebPage(website, index + 1);
  }
}

async function runWebsite(index) {
  if (websites && websites[index]) {
    const website = websites[index];
    await runWebPage(website, 0);
    await runWebsite(index + 1);
  }
}

async function run() {
  await runWebsite(0);
  SitespeedGrade.open(context, config);
  for (let index = 0; index < websites.length; index++) {
    const { type, name, pages } = websites[index];
    for (let index = 0; index < pages.length; index++) {
      const { name: page, url } = pages[index];
      const summaryPath = path.resolve(output, type, name, page, 'browsertime.pageSummary.json');
      if (await fse.exists(summaryPath)) {
        console.log(type, name, page);
        const summary = await fse.readJSON(summaryPath);
        SitespeedGrade.processGrade(summary);
        const { browserScripts, statistics: { custom: { grade } } } = summary;
        console.log(`    median: ${grade.median} | mean: ${grade.mean} | min: ${grade.min} | max: ${grade.max}`);
        for (let i = 0; i < browserScripts.length; i++) {
          const { custom: { grade: { metrics } } } = browserScripts[i];
          console.log(`    $${i}`);
          for (let j = 0; j < metrics.length; j++) {
            const { name, value, factor, grade } = metrics[j];
            console.log(`        ${name} - value: ${value} | factor: ${factor} | grade: ${grade}`);
          }
        }
      }
    }
  }
}

run();




test/latency.js
/**
 * 延迟测试
 */
const path = require('path');
const fs = require('fs');
const fse = require('fs-extra');
const shell = require('shelljs');
const chalk = require('chalk');
const findDataPath = require('./utils/findDataPath');

const context = path.resolve(__dirname, '../');
const output = path.resolve(context, 'output/latency');
const webpages = [
  {
    name: 'baidu',
    url: 'https://m.baidu.com'
  }
];
const latecies = [0, 100, 200, 300, 400, 500, 600, 700, 800, 900];

function test({ name, url, latency }) {
  return new Promise(async (resolve, reject) => {
    const outoutPath = path.resolve(output, name, String(latency));
    const cmd = `sitespeed.io ${url} --config ./.sitespeedio.test.json --browsertime.connectivity.latency ${latency} --browsertime.connectivity.upstreamKbps 9216 --browsertime.connectivity.downstreamKbps 9216 --outputFolder ${outoutPath}`;
    console.log('$', chalk.blue(cmd));
    if (await fse.exists(outoutPath)) {
      reject(new Error('aready run.'));
      return;
    }
    await fse.ensureDirSync(outoutPath);
    await fse.emptyDir(outoutPath);
    shell.exec(
      cmd,
      {
        cwd: context,
        async: true,
        fatal: false,
        silent: false
      },
      async (code, stdout, stderr) => {
        if (code === 0) {
          const pagesPath = path.resolve(outoutPath, 'pages');
          await fse.copy(await findDataPath(pagesPath), outoutPath, {
            overwrite: true
          });
          await fse.emptyDir(pagesPath);
          resolve(stdout);
        } else {
          reject(new Error(stderr));
        }
      }
    );
  });
}

async function runLatency(webpage, index) {
  if (index < latecies.length) {
    try {
      await test({
        name: webpage.name,
        url: webpage.url,
        latency: latecies[index]
      });
      console.log(chalk.green('success\n'));
    } catch (error) {
      console.log(chalk.red(`failure: ${error.message}\n`));
    }
    await runLatency(webpage, index + 1);
  }
}

async function runWebpage(index) {
  if (index < webpages.length) {
    const webpage = webpages[index];
    await runLatency(webpage, 0);
    await runWebpage(index + 1);
  }
}

async function run() {
  await runWebpage(0);
  for (let index = 0; index < webpages.length; index++) {
    const { name } = webpages[index];
    for (let index = 0; index < latecies.length; index++) {
      const latency = latecies[index];
      const summaryPath = path.resolve(
        output,
        name,
        String(latency),
        'browsertime.pageSummary.json'
      );
      if (await fse.exists(summaryPath)) {
        console.log(name, latency);
        const summary = await fse.readJSON(summaryPath);
        const {
          statistics: {
            timings: { pageTimings }
          }
        } = summary;
        printMetric('connect', pageTimings.serverConnectionTime);
        printMetric('response', pageTimings.serverResponseTime);
        printMetric('download', pageTimings.pageDownloadTime);
      }
    }
  }
}

function printMetric(name, metric) {
  console.log(
    ` ${name} - median: ${metric.median} | mean: ${metric.mean} | min: ${metric.min} | max: ${metric.max}`
  );
}

run();



test/download.js
/**
 * 延迟测试
 */
const path = require('path');
const fs = require('fs');
const fse = require('fs-extra');
const shell = require('shelljs');
const chalk = require('chalk');
const findDataPath = require('./utils/findDataPath');

const context = path.resolve(__dirname, '../');
const output = path.resolve(context, 'output/download');
const webpages = [
  {
    name: 'baidu',
    url: 'https://m.baidu.com'
  }
];
const downloads = [1024, 2048, 3072, 4096, 5120, 6144, 7168, 8192, 9216];

function test({ name, url, download }) {
  return new Promise(async (resolve, reject) => {
    const outoutPath = path.resolve(output, name, String(download));
    const cmd = `sitespeed.io ${url} --config ./.sitespeedio.test.json --browsertime.connectivity.latency 0 --browsertime.connectivity.upstreamKbps ${download / 4} --browsertime.connectivity.downstreamKbps ${download} --outputFolder ${outoutPath}`;
    console.log('$', chalk.blue(cmd));
    if (await fse.exists(outoutPath)) {
      reject(new Error('aready run.'));
      return;
    }
    await fse.ensureDirSync(outoutPath);
    await fse.emptyDir(outoutPath);
    shell.exec(
      cmd,
      {
        cwd: context,
        async: true,
        fatal: false,
        silent: false
      },
      async (code, stdout, stderr) => {
        if (code === 0) {
          const pagesPath = path.resolve(outoutPath, 'pages');
          await fse.copy(await findDataPath(pagesPath), outoutPath, {
            overwrite: true
          });
          await fse.emptyDir(pagesPath);
          resolve(stdout);
        } else {
          reject(new Error(stderr));
        }
      }
    );
  });
}

async function runDownload(webpage, index) {
  if (index < downloads.length) {
    try {
      await test({
        name: webpage.name,
        url: webpage.url,
        download: downloads[index]
      });
      console.log(chalk.green('success\n'));
    } catch (error) {
      console.log(chalk.red(`failure: ${error.message}\n`));
    }
    await runDownload(webpage, index + 1);
  }
}

async function runWebpage(index) {
  if (index < webpages.length) {
    const webpage = webpages[index];
    await runDownload(webpage, 0);
    await runWebpage(index + 1);
  }
}

async function run() {
  await runWebpage(0);
  for (let index = 0; index < webpages.length; index++) {
    const { name } = webpages[index];
    for (let index = 0; index < downloads.length; index++) {
      const download = downloads[index];
      const summaryPath = path.resolve(
        output,
        name,
        String(download),
        'browsertime.pageSummary.json'
      );
      if (await fse.exists(summaryPath)) {
        console.log(name, download);
        const summary = await fse.readJSON(summaryPath);
        const {
          statistics: {
            timings: { pageTimings }
          }
        } = summary;
        printMetric('connect', pageTimings.serverConnectionTime);
        printMetric('response', pageTimings.serverResponseTime);
        printMetric('download', pageTimings.pageDownloadTime);
      }
    }
  }
}

function printMetric(name, metric) {
  console.log(
    ` ${name} - median: ${metric.median} | mean: ${metric.mean} | min: ${metric.min} | max: ${metric.max}`
  );
}

run();



lib/index.js
const Stats = require('./fast-stats').Stats;
const statsHelpers = require('./statsHelpers');
const defaultConfig = require('./defaultConfig');

const getValue = (data, path) => {
  return path.split('.').reduce((rcc, item) => {
    if (rcc === undefined) {
      return undefined;
    }
    return rcc[item] !== undefined ? rcc[item] : undefined;
  }, data);
};

/**
 * @typedef Metric
 * @property {string} name 指标名称
 * @property {number} weight 指标权重
 * @property {boolean} negative 是否不计入总权重
 * @property {Array|Object} targets 指标目标值
 * @property {String|Array} value 指标取值
 */

/**
 * @typedef Grade
 * @property {string} name 指标名称
 * @property {number} weight 指标权重
 * @property {boolean} negative 是否不计入总权重
 * @property {number} value 指标实际值
 * @property {number} factor 指标取值因子
 * @property {number} grade 指标得分
 */

module.exports = {
  name() {
    return 'grade';
  },
  open(context, options) {
    this.network = options.browsertime.connectivity.alias;
    this.options = Object.assign({}, defaultConfig, options.grade);
  },
  processGrade(data) {
    const stats = new Stats();
    data.browserScripts.map((browserScript, index) => {
      /**
       * @type Grade[]
       */
      const grades = this.options.metrics.reduce((rcc, item) => {
        /**
         * @type Metric
         */
        const { name, weight, negative, targets, value } = item;

        // 确认目标值
        const target =
          Array.isArray(targets) ||
          typeof targets === 'number' ||
          typeof targets === 'function'
            ? targets
            : targets[this.network];
        if (target === undefined) {
          return rcc;
        }

        // 计算实际值
        let realValue;
        if (typeof value === 'string') {
          realValue = getValue(data, value.replace('${iteration}', index));
        } else if (Array.isArray(value) && value[0] && value[1]) {
          const valueA = getValue(
            data,
            value[0].replace('${iteration}', index)
          );
          const valueB = getValue(
            data,
            value[1].replace('${iteration}', index)
          );
          if (valueA !== undefined && valueB !== undefined && valueA > valueB) {
            realValue = valueA - valueB;
          }
        }

        // 计算得分
        if (realValue !== undefined && target !== undefined && realValue >= 0) {
          const result = {
            name,
            weight,
            negative,
            value: realValue
          };
          let matchedFactor;
          if (typeof target === 'function') {
            matchedFactor = target(realValue);
          } else if (Array.isArray(target)) {
            const matched = target.some((item, index) => {
              if (realValue <= item[0] || index === target.length - 1) {
                matchedFactor = item[1];
                return true;
              }
              return false;
            });
            if (!matched) {
              matchedFactor = 0;
            }
          } else {
            matchedFactor = Math.max(target / realValue, 1);
          }
          if (matchedFactor > 1) {
            matchedFactor = 1;
          }
          result.factor = matchedFactor;
          result.grade = this.options.total * matchedFactor;
          rcc.push(result);
        }
        return rcc;
      }, []);
      const totalWeight = grades.reduce((weight, item) => {
        return weight + (item.negative !== true ? item.weight : 0);
      }, 0);
      let finalGrade = 0;
      if (totalWeight > 0) {
        finalGrade = Number(
          grades
            .reduce((grade, item) => {
              return grade + item.grade * (item.weight / totalWeight);
            }, 0)
            .toFixed(1)
        );
      }
      if (finalGrade < 0) {
        finalGrade = 0;
      }
      browserScript.custom = Object.assign({}, browserScript.custom, {
        grade: {
          total: finalGrade,
          metrics: grades
        }
      });
      stats.push(finalGrade);
    });
    const statistics = statsHelpers.summarizeStats(stats, {
      decimals: 1
    });
    data.statistics = Object.assign(data.statistics, {
      custom: Object.assign(data.statistics.custom || {}, {
        grade: statistics
      })
    });
  },
  processMessage(message, queue) {
    switch (message.type) {
      case 'browsertime.pageSummary': {
        if (message.data && message.data.browserScripts) {
          this.processGrade(message.data);
        }
        break;
      }
      default:
        break;
    }
  }
};


lib/statsHelpers.js
'use strict';

const Stats = require('./fast-stats').Stats,
  get = require('./lodash.get'),
  set = require('./lodash.set');

function percentileName(percentile) {
  if (percentile === 0) {
    return 'min';
  } else if (percentile === 100) {
    return 'max';
  } else {
    return 'p' + String(percentile).replace('.', '_');
  }
}

function isFiniteNumber(n) {
  return typeof n === 'number' && isFinite(n);
}

module.exports = {
  /**
   * Create or update a fast-stats#Stats object in target at path.
   */
  pushStats(target, path, data) {
    if (!isFiniteNumber(data))
      throw new Error(`Tried to add ${data} to stats for path ${path}`);

    const stats = get(target, path, new Stats());
    stats.push(data);
    set(target, path, stats);
  },

  pushGroupStats(target, domainTarget, path, data) {
    this.pushStats(target, path, data);
    this.pushStats(domainTarget, path, data);
  },

  /**
   * Summarize stats and put result in target at path
   */
  setStatsSummary(target, path, stats) {
    set(target, path, this.summarizeStats(stats));
  },

  summarizeStats(stats, options) {
    if (stats.length === 0) {
      return undefined;
    }

    options = options || {};
    let percentiles = options.percentiles || [0, 90, 100];
    let decimals = options.decimals || 0;
    let data = {
      median: parseFloat(stats.median().toFixed(decimals)),
      mean: parseFloat(stats.amean().toFixed(decimals))
    };
    percentiles.forEach(p => {
      let name = percentileName(p);
      const percentile = stats.percentile(p);
      if (Number.isFinite(percentile)) {
        data[name] = parseFloat(percentile.toFixed(decimals));
      } else {
        throw new Error(
          'Failed to calculate ' +
            name +
            ' for stats: ' +
            JSON.stringify(stats, null, 2)
        );
      }
    });
    if (options.includeSum) {
      data.sum = parseFloat(stats.Σ().toFixed(decimals));
    }

    return data;
  }
};


lib/defaultConfig.js
/**
 * @example
 * ```js
 * var formula = make(1, 2);
 * formula(0); // 1
 * formula(1); // 0.25
 * ```
 * @example
 * ```js
 * var formula = make(10000, 2);
 * formula(0); // 1
 * formula(1000); // 0.826
 * formula(10000); // 0.25
 * ```
 */
function formula(scale, exponent) {
  return function (base) {
    return 1 / Math.pow(base / scale + 1, exponent)
  };
}

function testFormular(func) {
  var data = [
    100,
    200,
    400,
    600,
    800,
    1000,
    1200,
    1400,
    1500,
    1600,
    1800,
    2000,
    2500,
    3000,
    4000,
    5000,
    6000,
    7000,
    8000,
    9000,
    10000
  ];
  for (var index = 0; index < data.length; index++) {
    var time = data[index];
    console.log(time + ':', func(time));
  }
}

/**
 * htmlTransferSize - browserScripts.${iteration}.pageinfo.documentSize.transferSize
 * redirectionTime - browserScripts.${iteration}.timings.pageTimings.redirectionTime
 * serverConnectionTime - browserScripts.${iteration}.timings.pageTimings.serverConnectionTime
 * serverResponseTime - browserScripts.${iteration}.timings.pageTimings.serverResponseTime
 * pageDownloadTime - browserScripts.${iteration}.timings.pageTimings.pageDownloadTime
 * firstPaint - browserScripts.${iteration}.timings.paintTiming.first-paint
 * firstContentfulPaint - browserScripts.${iteration}.timings.paintTiming.first-contentful-paint
 * firstMeaningfulPaint - browserScripts.${iteration}.cdp.performance.FirstMeaningfulPaint
 * domContentLoadedTime - browserScripts.${iteration}.timings.navigationTiming.domContentLoadedEventEnd
 * pageLoadTime - browserScripts.${iteration}.timings.navigationTiming.loadEventEnd
 * speedIndex - browserScripts.${iteration}.visualMetrics.SpeedIndex
 */
const defaultConfig = {
  total: 10,
  metrics: [
    // 重定向
    {
      name: 'redirectionTime',
      weight: 10,
      negative: true,
      /**
       * @test testFormular(value => formula(1000, 2)(value) - 1)
       * 100: -0.17355371900826455
       * 200: -0.3055555555555556
       * 400: -0.4897959183673468
       * 600: -0.609375
       * 800: -0.691358024691358
       * 1000: -0.75
       * 1200: -0.7933884297520661
       * 1400: -0.8263888888888888
       * 1500: -0.84
       * 1600: -0.8520710059171598
       * 1800: -0.8724489795918366
       * 2000: -0.8888888888888888
       * 2500: -0.9183673469387755
       * 3000: -0.9375
       * 4000: -0.96
       * 5000: -0.9722222222222222
       * 6000: -0.9795918367346939
       * 7000: -0.984375
       * 8000: -0.9876543209876543
       * 9000: -0.99
       * 10000: -0.9917355371900827
       */
      targets: value => formula(1000, 2)(value) - 1,
      value: 'browserScripts.${iteration}.timings.pageTimings.redirectionTime'
    },
    // 首次绘制
    {
      name: 'firstPaint',
      weight: 5,
      /**
       * @test testFormular(formula(5000, 2))
       * 100: 0.9611687812379854
       * 200: 0.9245562130177514
       * 400: 0.8573388203017832
       * 600: 0.7971938775510203
       * 800: 0.7431629013079668
       * 1000: 0.6944444444444444
       * 1200: 0.6503642039542143
       * 1400: 0.6103515625
       * 1500: 0.5917159763313609
       * 1600: 0.573921028466483
       * 1800: 0.5406574394463669
       * 2000: 0.5102040816326532
       * 2500: 0.4444444444444444
       * 3000: 0.39062499999999994
       * 4000: 0.30864197530864196
       * 5000: 0.25
       * 6000: 0.20661157024793386
       * 7000: 0.1736111111111111
       * 8000: 0.14792899408284022
       * 9000: 0.1275510204081633
       * 10000: 0.1111111111111111
       */
      targets: formula(5000, 2),
      value: 'browserScripts.${iteration}.timings.paintTiming.first-paint'
    },
    // 首次有效绘制
    {
      name: 'firstMeaningfulPaint',
      weight: 10,
      /**
       * @test testFormular(formula(5000, 2))
       * 100: 0.9611687812379854
       * 200: 0.9245562130177514
       * 400: 0.8573388203017832
       * 600: 0.7971938775510203
       * 800: 0.7431629013079668
       * 1000: 0.6944444444444444
       * 1200: 0.6503642039542143
       * 1400: 0.6103515625
       * 1500: 0.5917159763313609
       * 1600: 0.573921028466483
       * 1800: 0.5406574394463669
       * 2000: 0.5102040816326532
       * 2500: 0.4444444444444444
       * 3000: 0.39062499999999994
       * 4000: 0.30864197530864196
       * 5000: 0.25
       * 6000: 0.20661157024793386
       * 7000: 0.1736111111111111
       * 8000: 0.14792899408284022
       * 9000: 0.1275510204081633
       * 10000: 0.1111111111111111
       */
      targets: formula(5000, 2),
      value: 'browserScripts.${iteration}.cdp.performance.FirstMeaningfulPaint'
    },
    // 页面加载时长
    {
      name: 'pageLoadTime',
      weight: 5,
      /**
       * @test testFormular(value => formula(10000, 2)(Math.max(0, value - 1000)))
       * 100: 1
       * 200: 1
       * 400: 1
       * 600: 1
       * 800: 1
       * 1000: 1
       * 1200: 0.9611687812379854
       * 1400: 0.9245562130177514
       * 1500: 0.9070294784580498
       * 1600: 0.8899964400142398
       * 1800: 0.8573388203017832
       * 2000: 0.8264462809917354
       * 2500: 0.7561436672967865
       * 3000: 0.6944444444444444
       * 4000: 0.5917159763313609
       * 5000: 0.5102040816326532
       * 6000: 0.4444444444444444
       * 7000: 0.39062499999999994
       * 8000: 0.34602076124567477
       * 9000: 0.30864197530864196
       * 10000: 0.2770083102493075
       */
      targets: value => formula(10000, 2)(Math.max(0, value - 1000)),
      value: 'browserScripts.${iteration}.timings.navigationTiming.loadEventEnd'
    },
    // 速度指数
    {
      name: 'speedIndex',
      weight: 8,
      /**
       * @test testFormular(value => formula(10000, 2)(Math.max(0, value - 1000)))
       * 100: 1
       * 200: 1
       * 400: 1
       * 600: 1
       * 800: 1
       * 1000: 1
       * 1200: 0.9611687812379854
       * 1400: 0.9245562130177514
       * 1500: 0.9070294784580498
       * 1600: 0.8899964400142398
       * 1800: 0.8573388203017832
       * 2000: 0.8264462809917354
       * 2500: 0.7561436672967865
       * 3000: 0.6944444444444444
       * 4000: 0.5917159763313609
       * 5000: 0.5102040816326532
       * 6000: 0.4444444444444444
       * 7000: 0.39062499999999994
       * 8000: 0.34602076124567477
       * 9000: 0.30864197530864196
       * 10000: 0.2770083102493075
       */
      targets: value => formula(10000, 2)(Math.max(0, value - 1000)),
      value: 'browserScripts.${iteration}.visualMetrics.SpeedIndex'
    }
  ]
};

module.exports = defaultConfig;
